### PR TITLE
Dsdl compiler v1.0

### PR DESCRIPTION
**Pull request by aasmune**
This PR updates the DSDL compiler for Libcanard from v0.1 to v1.0. The work on this PR was started before Scott started updating the DSDL compiler for Libuavcan. However, we both chose similar approaches. 

The biggest change introduced here is the (natural) change from using pyuavcan to pydsdl to parse the DSDL. Please note that this PR is blocked by UAVCAN/pydsdl#16.

As I understood from @kjetilkjeka, signatures will not we used in v1.0, and this is therefore removed from the template for the C header/source files.

To verify the behavior, I updated the DSDL for v0.1 to include version numbering, and temporarily changed the port ID ranges for standard messages/services to be able to parse v0.1 DSDL with pydsdl. The result can be seen in the attached file, which shows that the result is identical, beside the removed signatures, and the changes needed to adapt v0.1 DSDL to be able to use pydsdl.
[compare_dsdl_v0.1_v1.0.txt](https://github.com/UAVCAN/libcanard/files/2873383/compare_dsdl_v0.1_v1.0.txt)

### CODE REVIEW

File: dsdl_compiler/libcanard_dsdl_compiler/__init__.py [hunk 1]
```diff
@@ -305,46 +318,42 @@ def inject_cpp_types(attributes):
...
-    if t.kind == t.KIND_MESSAGE:
+    if isinstance(t, (StructureType, UnionType)):
         t.has_array = inject_cpp_types(t.fields)
         t.has_float16 = has_float16(t.fields)
         inject_cpp_types(t.constants)
         t.all_attributes = t.fields + t.constants
-        t.union = t.union and len(t.fields)
+        t.union = isinstance(t, UnionType)
         if t.union:
             t.union = len(t.fields).bit_length()
```
Comment (pavel-kirienko) [inline dsdl_compiler/libcanard_dsdl_compiler/__init__.py]: Will break if `len(t.fields)` is a power of two. The correct expression is `(len(t.fields) - 1).bit_length()`. Per the spec, a union cannot contain less than 2 fields, this is enforced by PyDSDL.

File: dsdl_compiler/libcanard_dsdl_compiler/__init__.py [hunk 2]
```diff
@@ -305,46 +318,42 @@ def inject_cpp_types(attributes):
...
+        inject_cpp_types(t.request_type.constants)
+        t.response_has_array = inject_cpp_types(t.response_type.fields)
+        t.response_has_float16 = has_float16(t.response_type.fields)
+        inject_cpp_types(t.response_type.constants)
+        t.all_attributes = t.request_type.fields + t.request_type.constants + t.response_type.fields + t.response_type.constants
+        t.request_union = isinstance(t.request_type, UnionType) and len(t.request_type.fields)
+        t.response_union = isinstance(t.response_type, UnionType) and len(t.response_fields)
         if t.request_union:
             t.request_union = len(t.request_fields).bit_length()
         if t.response_union:
```
Comment (pavel-kirienko) [inline dsdl_compiler/libcanard_dsdl_compiler/__init__.py]: Same here, the expression is incorrect, see above.

**Comment by aasmune** [comment]
> I predict that this transpiler will fail to produce a meaningful output when there are multiple versions of the same data type available. The version numbers could be encoded in the file name, e.g., `Heartbeat.1.0.h` (interestingly, dot separators seem to be permitted in header file name both by the standard and by MISRA). Correct version numbers must be specified when referring to a nested type name, too. 

If I understood you correctly, version numbers should be added to the file names, but not to the generated values/types in the files?

> For ease of application maintenance, it might make sense to also generate symlinks to the newest minor version, e.g. `Foo.1.newest.h` (not sure about `newest`) pointing to `Foo.1.2.h`. 

I like the idea of generating a symlink which will always point to the newest minor version, I will look into it!

**Comment by pavel-kirienko** [comment]
> If I understood you correctly, version numbers should be added to the file names, but not to the generated values/types in the files?

Both; I should have been more specific. Otherwise, there might be ODR violations if several versions are used concurrently. I see now that it's going to get clumsy fast, since there aren't any decent means of scoping available in C, but oh well, it's C.

**Comment by aasmune** [comment]
> > If I understood you correctly, version numbers should be added to the file names, but not to the generated values/types in the files?
> 
> Both; I should have been more specific. Otherwise, there might be ODR violations if several versions are used concurrently. I see now that it's going to get clumsy fast, since there aren't any decent means of scoping available in C, but oh well, it's C.

If this is the case, I don't see the point in having a symlink with _newest_ as the code also will need to be updated each time a new minor version is published as the previous version number is "hard coded" in application code. 

How would you recommend include version numbers in the actual code? Just add it at the end e.g. `uavcan_node_Heartbeat.1.0` and `uavcan_node_Heartbeat.1.1`?

**Comment by pavel-kirienko** [comment]
> If this is the case, I don't see the point in having a symlink with _newest_ as the code also will need to be updated each time a new minor version is published as the previous version number is "hard coded" in application code.

You are correct; in my original post, I failed to consider the ODR-related issues. The need for code update could be mitigated by replacing the symlink with a dedicated header containing an alias to the latest data type (including its serialization functions, defines, etc). Not sure if it's worth the effort at this stage.

> How would you recommend include version numbers in the actual code? Just add it at the end e.g. `uavcan_node_Heartbeat.1.0` and `uavcan_node_Heartbeat.1.1`?

Yep, I don't see any other option here, it's C (surely you meant `uavcan_node_Heartbeat_1_0`).

**Comment by aasmune** [comment]
> > If this is the case, I don't see the point in having a symlink with _newest_ as the code also will need to be updated each time a new minor version is published as the previous version number is "hard coded" in application code.
> 
> You are correct; in my original post, I failed to consider the ODR-related issues. The need for code update could be mitigated by replacing the symlink with a dedicated header containing an alias to the latest data type (including its serialization functions, defines, etc). Not sure if it's worth the effort at this stage.

Could it be an idea to use the name without any versions as an alias for the latest data type? This means that applications will be able to always use the latest version without explicitly updating all code if they want, or they can "freeze" at a specific version by using the types including version numbers.

By "alias" you mean `typedef`?

> > How would you recommend include version numbers in the actual code? Just add it at the end e.g. `uavcan_node_Heartbeat.1.0` and `uavcan_node_Heartbeat.1.1`?
> 
> Yep, I don't see any other option here, it's C (surely you meant `uavcan_node_Heartbeat_1_0`).

Of course!

**Comment by pavel-kirienko** [comment]
> Could it be an idea to use the name without any versions as an alias for the latest data type?

Auto-update across major versions is of no practical value, since that pretty much defeats the wire compatibility guarantees provided by DSDL. Minor versions are different, since their differences are not manifested on the wire in any way.

**Comment by aasmune** [comment]
> > Could it be an idea to use the name without any versions as an alias for the latest data type?
> 
> Auto-update across major versions is of no practical value, since that pretty much defeats the wire compatibility guarantees provided by DSDL. Minor versions are different, since their differences are not manifested on the wire in any way.

I see. So if this should be done, it should be restricted to the latest minor version for each major. I will have a look at it, but this is also something which could possibly be postponed to another time as it is not crucial for matching the v.1.0 specification, and is, therefore, more of a "nice-to-have" feature than a "must-have" in my opinion.

**Comment by pavel-kirienko** [comment]
> this is also something which could possibly be postponed to another time
as it is not crucial for matching the v.1.0 specification, and is,
therefore, more of a "nice-to-have" feature than a "must-have" in my
opinion.

Yes, certainly, this is nowhere critical atm.

**Comment by aasmune** [comment]
I have now added versions to all generated code + file names + fixed most other review feedback. Will look more into how much work/the possibilities of creating aliases hopefully tomorrow

**Pull request merged**